### ISSUE TITLE
Running `NuGet.exe pack` in parallel causes System.ObjectDisposedException

### ISSUE THREAD

**Issue by deepakaravindr** [NuGet/Home#1282]
@zentron

Moved from https://github.com/NuGet/NuGet.CommandLine/issues/18

Running the following pack script against several different packages (note completely different source and destination)

```
 cls

  $ScriptBlock = {
    param($ver) 
    Write-Host "[processing Package # '$ver' inside the job]"
    & "C:\Temp\NuGet.exe" pack "D:\Development\Packages\Acme.Web$ver\Acme.Web.nuspec" -NoPackageAnalysis -BasePath "D:\Development\Packages\Acme.Web$ver" -OutputDirectory "C:\Temp" -Version 1.0.$ver -Verbosity Detailed 
  }

#Run 3 Jobs
  for($i=1; $i -le 3; $i++) {
    Start-Job $ScriptBlock -ArgumentList $i
  }

 # Wait for all to complete
While (Get-Job -State "Running") { Start-Sleep 2 }

# Display output from all jobs
Get-Job | Receive-Job

Remove-Job *
```

Results in 2/3 failing with 

```
System.ObjectDisposedException: Can not access a closed Stream.
    + CategoryInfo          : NotSpecified: (System.ObjectDi... closed Stream.:String) [], Remote 
   Exception
    + FullyQualifiedErrorId : NativeCommandError
    + PSComputerName        : localhost

   at System.IO.Compression.DeflateStream.EnsureNotDisposed()
   at System.IO.Compression.DeflateStream.Flush()
   at MS.Internal.IO.Packaging.CompressStream.Flush()
   at MS.Internal.IO.Z
ip.ProgressiveCrcCalculatingStream.Flush()
   at MS.Internal.IO.Zip.ZipIOLocalFileBlock.FlushExposedStreams()
   at MS.Internal.IO.Zip.ZipIOLocalFileBlock.UpdateReferences(Boolean closingFlag)
   at MS.Internal.IO.Zip.ZipIOBlockManager.SaveContainer(Boolean closingFlag)
   at MS.Internal.IO.Zip.ZipIOBlockManager.SaveStream(ZipIOLocalFileBlock blockRequestingFlush, 
Boolean closingFlag)
   at MS.Internal.IO.Zip.ZipIOModeEnforcingStream.Dispose(Boolean disposing)
   at System.IO.Stream.Close()
   at System.IO.Packaging.PackagePart.Close()
   at System.IO.Packaging.Package.DoClose(PackagePart p)
   at System.IO.Packaging.Package.DoOperationOnEachPart(PartOperation operation)
   at System.IO.Packaging.Package.System.IDisposable.Dispose()
   at NuGet.PackageBuilder.Save(Stream stream)
   at NuGet.Commands.PackCommand.BuildPackage(PackageBuilder builder, String outputPath)
   at NuGet.Commands.PackCommand.BuildFromNuspec(String path)
   at NuGet.Commands.PackCommand.BuildPackage(String path)
   at NuGet.Commands.PackCommand.ExecuteCommand()
   at NuGet.Commands.Command.Execute()
   at NuGet.Program.Main(String[] args)
```

Not sure if there is some limitation or argument not being supplied to get this process to work.

**Comment by deepakaravindr** [comment]
@zentron referenced this issue in OctopusDeploy/Issues 11 days ago
 Open
Octopack cannot run in parallel: OctoPack error OCTONUGET: Can not access a closed Stream #1684

**Comment by BertusGreeff** [comment]
+1 Also hitting this in our build system thats using NuProj to generate multiple packages in parralel.

**Comment by joel74** [comment]
+1. I'm not able to upgrade to newer than 2.8.1 without running into this issue when packing in parallel.

**Comment by MartinRykfors** [comment]
I was getting the same error when trying to make a nuget package containing a very large file.

I did not have enough space on my C: drive, so it seems that the compression stage failed because there was not enough space for any temp files needed.

I'm not sure this applies to everyone of you who have faced this problem, but freeing up space on C: solved the problem for me.

**Comment by mausch** [comment]
I've been [having this issue since August](https://twitter.com/mausch/statuses/633293648340348928) (at least) as well. It is definitely not due to lack of disk space.
This bug is especially annoying because `pack` is really slow for some reason, and parallelising it seemed like a cheap work around.

**Comment by emgarten** [comment]
@zentron @mausch @joel74 @BertusGreeff  I'm not able to repro this issue. I've tried the repro script in this bug and also bumped it up to packing 10 packages at the same time. This along with increasing the size of the packages and using older versions of NuGet.exe such as 2.8.6 had no effect. The packages were created in parallel without errors each time.

Would you be able to provide more details on reproducing this or set up a machine where I can debug this?

What version of .NET do you have the machines where you are seeing this and what version of windows is it? What are the machine specs?

**Comment by mausch** [comment]
I've seen it happen both on my desktop (.NET 4.5.2, Windows 8.1, Core i7, SSD, more than enough memory) and our build machine (EC2 t2.medium, .NET 4.5.2, Windows 8)

**Comment by emgarten** [comment]
@mausch how consistently does it happen for you?

**Comment by mausch** [comment]
I've disabled parallel packing several months ago... IIRC it was happening in around 20% of our builds. Not a good time right now to mess with our builds to try to reproduce it again but I'll try to create a smaller repro...

**Comment by IntelligentLabs** [comment]
To add to this if it helps, I get the exact same error if I run the NuGet.PackageBuilder.Save command via the Nuget.Core.dll on a large package via TFS build. If I run the command directly on the server I do not get the problem.

**Comment by mvalipour** [comment]
I have exact same issue in OctopusDeploy. My scenario is that my TeamCity kicks off multiple deployments (in fact releases that have an auto-deploy lifecycle), and in consequently tries to `pack` several packages at the same time (`nuget pack` I believe) which results in All of the 5 except last one to fail -- retrying each of 4 failed ones will result in success which is clear indication that problem is in `nuget pack` parallel processes interrupting each other.

Does anyone know any workaround for this?

**Comment by emgarten** [comment]
This has been fixed in NuGet.exe 3.4.0.
NuGet/NuGet.Client@113f7d3697bb89c9b2ad077941391f9cba5f7d2f

**Comment by alfnilsson** [comment]
I'm still getting this in NuGet 3.4.2.830 when running multiple nuget pack at the same time.

2016-04-10T17:16:36.3373767Z Attempting to build package from 'NetR.Project.3.1604.0.1604101916.nuspec'.
2016-04-10T17:17:48.7852510Z NetR.Project\modules_protected\Shell\Shell.zip
2016-04-10T17:17:48.9884099Z ##[error]Can not access a closed Stream.

The filename referring to is a quite large file at 9.85 MB

**Comment by emgarten** [comment]
@alfnilsson this like a different problem, would you try running with `-verbosity detailed` to get a stack trace for this failure and then log a new issue?
